src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -190,14 +190,15 @@ bool bindDisposable(bool fromExpression, out MethodArgumentInfo? patternDisposeI
 
                 // Pattern-based binding
                 // If this is a ref struct, or we're in a valid asynchronous using, try binding via pattern.
+                BindingDiagnosticBag? patternDiagnostics = null;
```

Comment:
>BindingDiagnosticBag? patternDiagnostics = null; [](http://example.com/codeflow?start=16&length=48)

I find this change unnecessary complicated and it feels like we can simply follow what ```GetDisposalInfoForEnumerator``` is doing in terms of diagnostics. Specifically, it looks like it reports ```patternDiagnostics``` only when a pattern method is found. Otherwise, we either find an interface implementation, or report an error about that if we don't.

Reply:
> Specifically, it looks like it reports patternDiagnostics only when a pattern method is found.

That's not what I see in `GetDisposalInfoForEnumerator` (we always discard there): `TryFindDisposePatternMethod(receiver, syntax, isAsync, BindingDiagnosticBag.Discarded, out bool expanded)`

The options so far:
1. always discard pattern-based diagnostics
2. report pattern-based diagnostics when we commit to pattern method
3. report pattern-based diagnostics when we commit to pattern method or when interface fallback fails

I don't see what's wrong with option 3.

If we go with option 2 as you're describing above then we lose the accessibility diagnostic in this scenario, which seems a useful diagnostic:
```
        public void TestPatternBasedDisposal_InstanceMethod_Inaccessible()
        {
            string source = @"
public class D
{
    public static async System.Threading.Tasks.Task<int> Main()
    {
        await using (var x = new C())
        {
        }

        return 1;
    }
}
public class C
{
    private System.Threading.Tasks.ValueTask DisposeAsync()
        => throw null;
}
";
            var comp = CreateCompilationWithTasksExtensions(new[] { source, IAsyncDisposableDefinition });
            comp.VerifyDiagnostics(
                // (6,22): error CS0122: 'C.DisposeAsync()' is inaccessible due to its protection level
                //         await using (var x = new C())
                Diagnostic(ErrorCode.ERR_BadAccess, "var x = new C()").WithArguments("C.DisposeAsync()").WithLocation(6, 22),
                // (6,22): error CS8410: 'C': type used in an async using statement must be implicitly convertible to 'System.IAsyncDisposable' or implement a suitable 'DisposeAsync' method.
                //         await using (var x = new C())
                Diagnostic(ErrorCode.ERR_NoConvToIAsyncDisp, "var x = new C()").WithArguments("C").WithLocation(6, 22)
                );
        }
```

If we go with option 1 as `GetDisposalInfoForEnumerator` does, then we furthermore lose the obsolete diagnostic in this scenario, which also seems a good diagnostic:
```
        public void TestWithObsoleteDisposeAsync_02()
        {
            string source = @"
class C : System.IAsyncDisposable
{
    public static async System.Threading.Tasks.Task Main()
    {
        await using (var x = new C())
        {
        }
    }
    [System.Obsolete]
    public async System.Threading.Tasks.ValueTask DisposeAsync()
    {
        await System.Threading.Tasks.Task.Yield();
    }
}
";
            var comp = CreateCompilationWithTasksExtensions(new[] { source, IAsyncDisposableDefinition }, options: TestOptions.DebugExe);
            comp.VerifyDiagnostics(
                // 0.cs(6,22): warning CS0612: 'C.DisposeAsync()' is obsolete
                //         await using (var x = new C())
                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "var x = new C()").WithArguments("C.DisposeAsync()").WithLocation(6, 22)
                );
        }
```

Reply:
From offline discussion, we'll go with option 2.
There's already a tracking issue for `GetDisposalInfoForEnumerator` dropping obsolete diagnostics: [https://github.com/dotnet/roslyn/issues/30257](https://github.com/dotnet/roslyn/issues/30257)